src/lib/core/exception.h

```diff
@@ -231,14 +292,42 @@ class CryptoError: public Exception {
 	}
 
 	virtual void raise() { throw this; }
+
+	virtual struct error *dup() const
+	{
+		return dup_error<CryptoError>(this);
+	}
 };
 
 class RaftError: public Exception {
 public:
+	RaftError() : Exception(&type_RaftError, NULL, 0) {}
+
 	RaftError(const char *file, unsigned line, const char *format, ...);
 	virtual void raise() { throw this; }
+	virtual struct error *dup() const
+	{
+		return dup_error<RaftError>(this);
+	}
 };
 
+template <class ErrorT>
+struct error *
+Exception::dup_error(const struct error *to_dup)
+{
+	try {
+		struct error *e = new ErrorT;
+		memcpy(e, to_dup, sizeof(*e));
+		e->refs = 0;
+		error_payload_copy(&e->payload, &to_dup->payload);
+		e->cause = NULL;
+		e->effect = NULL;
+		return e;
+	} catch (OutOfMemory *ex) {
+		return ex;
```

Comment:
You say that this function 'Duplicates an error with the template type'. But in this line you silently return another error. That's a very strange behavior.

Reply:
I'l move try catch block to [`exception_dup`](https://github.com/KaitmazianMO/tarantool/blob/efb1102d9f404cf49cc44296f08432b8ff1268c2/src/lib/core/exception.cc#L62-L67) function. In case of memory allocation error this function will return an `OutOfMemory` error becasue it can be called from C code wich doesn't support exceptions.

Reply:
Done. Exception can be raised while copying, so I changed `OutOfMemory` to the base class 'Exception'.

Reply:
There is no this function anymore.